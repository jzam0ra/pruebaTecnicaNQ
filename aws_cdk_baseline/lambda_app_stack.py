# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0
from typing import Dict
from os import path
from aws_cdk import (
    Stack,
    aws_lambda as lambda_,
    aws_s3_deployment as s3deploy,
    RemovalPolicy,
    CfnTag,
)
from constructs import Construct

class LambdaAppStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, config:Dict, stage:str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        ##########################################################################
        ##                                                                      ##
        ##                INSERTAR ARCHIVOS .PY EN S3 BUCKETS                   ##
        ##                                                                      ##
        ##                                                                      ##
        ##########################################################################

        self.lambda_bucket = s3.Bucket.from_bucket_arn(self, "lambda-Bucket",
            bucket_arn  = f"arn:aws:s3:::s3-scripts-bucket-{str(config[stage]['awsAccountId'])}"
        )

        self.bucket_deployment = s3deploy.BucketDeployment(self, "Populate-lambda-Bucket",
            sources=[s3deploy.Source.asset(
                path.join(path.dirname(__file__), "project_files/lambda"))],
            destination_bucket=self.lambda_bucket,
            prune = False, 
        )
         
        ##########################################################################
        ##                                                                      ##
        ##                 CREACION DE AWS CDK LAMBDA FUNCTIONS                 ##
        ##                                                                      ##
        ##                                                                      ##
        ##########################################################################

        self.lam_data_extraction = lambda_.CfnFunction(self, "lam_data_extraction",
            code=lambda_.CfnFunction.CodeProperty(
                s3_bucket=f"s3-scripts-bucket-{str(config[stage]['awsAccountId'])}",
                s3_key="lambda/lam_data_extraction.zip",
            ),
            ## este rol de servicio se podría definir con IaC pero acá se creó desde la consola y se pone el arn
            role=f"arn:aws:iam::{str(config[stage]['awsAccountId'])}:role/service-role/first-test-role-vn00mm7y", 
            # the properties below are optional
            architectures=["x86_64"], ###
            description="",
            ephemeral_storage=lambda_.CfnFunction.EphemeralStorageProperty(
                size=1024
            ),
            function_name=f"lam_data_extraction",
            layers=[f"arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15"],
            handler="lambda_function.lambda_handler",
            memory_size=1024,
            runtime="python3.12",
            tags=[CfnTag(
                key="Name",
                value=f"lam_data_extraction"
            )],
            timeout=300
        ) 
        self.lam_data_extraction.apply_removal_policy(RemovalPolicy.RETAIN)
        self.lam_data_extraction.node.add_dependency(self.bucket_deployment)

        self.lam_insert_data_catalog = lambda_.CfnFunction(self, "lam_insert_data_catalog",
            code=lambda_.CfnFunction.CodeProperty(
                s3_bucket=f"s3-scripts-bucket-{str(config[stage]['awsAccountId'])}",
                s3_key="lambda/lam_insert_data_catalog.zip",
            ),
            role=f"arn:aws:iam::{str(config[stage]['awsAccountId'])}:role/service-role/first-test-role-vn00mm7y",
            # the properties below are optional
            architectures=["x86_64"], ###
            description="",
            ephemeral_storage=lambda_.CfnFunction.EphemeralStorageProperty(
                size=1024
            ),
            function_name=f"lam_insert_data_catalog",
            layers=[f"arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15"],
            handler="lambda_function.lambda_handler",
            memory_size=1024,
            runtime="python3.12",
            tags=[CfnTag(
                key="Name",
                value=f"lam_insert_data_catalog"
            )],
            timeout=300
        ) 
        self.lam_insert_data_catalog.apply_removal_policy(RemovalPolicy.RETAIN)
        self.lam_insert_data_catalog.node.add_dependency(self.bucket_deployment)

        self.lam_transformar_precipitaciones = lambda_.CfnFunction(self, "lam_transformar_precipitaciones",
            code=lambda_.CfnFunction.CodeProperty(
                s3_bucket=f"s3-scripts-bucket-{str(config[stage]['awsAccountId'])}",
                s3_key="lambda/lam_transformar_precipitaciones.zip",
            ),
            role=f"arn:aws:iam::{str(config[stage]['awsAccountId'])}:role/service-role/first-test-role-vn00mm7y",
            # the properties below are optional
            architectures=["x86_64"], ###
            description="",
            ephemeral_storage=lambda_.CfnFunction.EphemeralStorageProperty(
                size=1024
            ),
            function_name=f"lam_transformar_precipitaciones",
            layers=[f"arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15"],
            handler="lambda_function.lambda_handler",
            memory_size=1024,
            runtime="python3.12",
            tags=[CfnTag(
                key="Name",
                value=f"lam_transformar_precipitaciones"
            )],
            timeout=300
        ) 
        self.lam_transformar_precipitaciones.apply_removal_policy(RemovalPolicy.RETAIN)
        self.lam_transformar_precipitaciones.node.add_dependency(self.bucket_deployment)