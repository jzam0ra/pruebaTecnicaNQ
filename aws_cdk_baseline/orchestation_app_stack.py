from typing import Dict
from os import path
from aws_cdk import (
    Stack,
    aws_stepfunctions as stepfunctions,
    aws_scheduler as scheduler,
    RemovalPolicy,
)
from constructs import Construct
import json


class OrchestationAppStack(Stack):
    def __init__(self, scope: Construct, construct_id: str, config:Dict, stage:str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)



        ##########################################################################
        ##                                                                      ##
        ##                   CREACION DE AWS STEP FUNCTIONS                     ##
        ##                                                                      ##
        ##                                                                      ##
        #########################################################################

        self.definition_stm = {
            "QueryLanguage": "JSONata",
            "Comment": "A description of my state machine",
            "StartAt": "Lambda Invoke",
            "States": {
                "Lambda Invoke": {
                    "Type": "Task",
                    "Resource": "arn:aws:states:::lambda:invoke",
                    "Output": "{% $states.result.Payload %}",
                    "Arguments": {
                        "FunctionName": "arn:aws:lambda:us-east-1:390403866377:function:first-test:$LATEST"
                    },
                    "Retry": [
                        {
                            "ErrorEquals": [
                                "Lambda.ServiceException",
                                "Lambda.AWSLambdaException",
                                "Lambda.SdkClientException",
                                "Lambda.TooManyRequestsException"
                            ],
                            "IntervalSeconds": 1,
                            "MaxAttempts": 3,
                            "BackoffRate": 2,
                            "JitterStrategy": "FULL"
                        }
                    ],
                    "End": True
                }
            }
        }


        self.stm_launch_lambda = stepfunctions.CfnStateMachine(self, "stm_launch_lambda",
            role_arn=f"arn:aws:iam::{str(config[stage]['awsAccountId'])}:role/stepfunctionsrole",
            state_machine_name=f"stm_launch_lambda",
            state_machine_type="STANDARD",
            # the properties below are optional
            definition_string=json.dumps(self.definition_stm),
            tags=[stepfunctions.CfnStateMachine.TagsEntryProperty(
                key="Name",
                value=f"stm_launch_lambda"
            )],
        )
        self.stm_launch_lambda.apply_removal_policy(RemovalPolicy.RETAIN)


        ##########################################################################
        ##                                                                      ##
        ##                   CREACION DE AWS EVENTBRIDGE RULES                  ##
        ##                             Y SCHEDULERS                             ##
        ##                                                                      ##
        ##########################################################################
        self.schedule_input = {}
        self.import_schedule = scheduler.CfnSchedule(self, "import_schedule",
            flexible_time_window=scheduler.CfnSchedule.FlexibleTimeWindowProperty(
                mode="OFF",
            ),
            schedule_expression="cron(0 9 * * ? *)",
            target=scheduler.CfnSchedule.TargetProperty(
                arn=f"arn:aws:states:us-east-1:{str(config[stage]['awsAccountId'])}:stateMachine:stm_launch_lambda",
                role_arn=f"arn:aws:iam::{str(config[stage]['awsAccountId'])}:role/service-role/Amazon_EventBridge_Scheduler_SFN_b61b2be1a1",
                input=json.dumps(self.schedule_input),
                retry_policy=scheduler.CfnSchedule.RetryPolicyProperty(
                    maximum_event_age_in_seconds=900,
                    maximum_retry_attempts=0
                )
            ),

            description="",
            group_name="default",
            name="import_schedule",
            schedule_expression_timezone="America/Bogota",
            state="DISABLED"
        )
        self.import_schedule.apply_removal_policy(RemovalPolicy.RETAIN)

        