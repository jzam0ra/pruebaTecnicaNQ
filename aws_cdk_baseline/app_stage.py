# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0
from typing import Dict
import aws_cdk as cdk
from constructs import Construct
from aws_cdk_baseline.lambda_app_stack import LambdaAppStack

class GlueAppStage(cdk.Stage):
    def __init__(self, scope: Construct, construct_id: str, config:Dict, stage:str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.lambda_app_stack = LambdaAppStack(self, "LambdaAppStack", config, stage)
