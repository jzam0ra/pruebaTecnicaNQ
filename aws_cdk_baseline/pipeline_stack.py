# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0
from typing import Dict
from aws_cdk import (
    Environment,
    Stack,
    aws_codecommit as codecommit,
    aws_iam as iam
)
from constructs import Construct
from aws_cdk.pipelines import CodePipeline, CodePipelineSource, CodeBuildStep, ManualApprovalStep, ShellStep
from helper import create_archive
from aws_cdk_baseline.app_stage import GlueAppStage


class PipelineStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, config: Dict, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        archive_file = create_archive()

        source = CodePipelineSource.connection("jzam0ra/pruebaTecnicaNQ", "main",
            connection_arn="arn:aws:codeconnections:us-east-1:390403866377:connection/3df4df9e-0eb1-4ffa-9979-8f61dbd46ade"
        )
        
        pipeline = CodePipeline(self, "DataPipeline",
            pipeline_name="DataPipeline",
            cross_account_keys=True,
            docker_enabled_for_synth=True,
            synth=CodeBuildStep("CdkSynth_UnitTest",
                input=source,
                install_commands=[
                    "pip install -r requirements-dev.txt",
                    "pip install -r requirements.txt",
                    "npm install -g aws-cdk",
                ],
                commands=[
                    "cdk synth -c stage=dev",
                    # Unit test for CDK stack
                    "python -m pytest",
                ],
            )
        )
        
        # Dev deployment
        dev_stage_name = "DeployDev"
        if config["devAccount"]["awsAccountId"] == config["devAccount"]["awsAccountId"] and config["devAccount"]["awsRegion"] == config["devAccount"]["awsRegion"]:
            dev_env = None
        else:
            dev_env = Environment(
                account=str(config["devAccount"]["awsAccountId"]), 
                region=config["devAccount"]["awsRegion"]
            )
        dev_stage_app = GlueAppStage(
            self, 
            dev_stage_name,
            config=config,
            stage="dev",
            env=dev_env
        )
        dev_stage = pipeline.add_stage(dev_stage_app)
        dev_stage.add_pre(ManualApprovalStep("Approval"))