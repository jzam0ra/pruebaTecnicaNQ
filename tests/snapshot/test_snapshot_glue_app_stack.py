# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0
import aws_cdk as core
import aws_cdk.assertions as assertions

from aws_cdk_baseline.lambda_app_stack import LambdaAppStack

config = {
    "devAccount": {
        "awsAccountId": 390403866377,
        "awsRegion": "us-east-1"
    },
    "dev": {
        "awsAccountId": 390403866377
    }
}

def remove_key(obj, key_to_remove):
    if isinstance(obj, dict):
        obj.pop(key_to_remove, None)
        for k, v in obj.items():
            remove_key(v, key_to_remove)
    elif isinstance(obj, list):
        for item in obj:
            remove_key(item, key_to_remove)


def test_glue_app_stack_snapshot(snapshot):
    app = core.App()
    stack = LambdaAppStack(
        app, 
        "TestGlueAppStack",
        config=config,    
        stage="dev",           
        env=core.Environment(region='us-east-1')
    )
    template = assertions.Template.from_stack(stack)
    templateJson = template.to_json() 
    remove_key(templateJson, "S3Key")
    remove_key(snapshot, "S3Key")
    assert templateJson == snapshot
