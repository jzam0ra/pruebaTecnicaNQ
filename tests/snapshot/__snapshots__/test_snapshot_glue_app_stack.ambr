# serializer version: 1
# name: test_glue_app_stack_snapshot
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536': dict({
        'DependsOn': list([
          'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF',
          'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265',
        ]),
        'Properties': dict({
          'Code': dict({
            'S3Bucket': dict({
              'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1',
            }),
          }),
          'Environment': dict({
            'Variables': dict({
              'AWS_CA_BUNDLE': '/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem',
            }),
          }),
          'Handler': 'index.handler',
          'Layers': list([
            dict({
              'Ref': 'PopulatelambdaBucketAwsCliLayer9FD2CBCF',
            }),
          ]),
          'Role': dict({
            'Fn::GetAtt': list([
              'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265',
              'Arn',
            ]),
          }),
          'Runtime': 'python3.9',
          'Timeout': 900,
        }),
        'Type': 'AWS::Lambda::Function',
      }),
      'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265': dict({
        'Properties': dict({
          'AssumeRolePolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'sts:AssumeRole',
                'Effect': 'Allow',
                'Principal': dict({
                  'Service': 'lambda.amazonaws.com',
                }),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'ManagedPolicyArns': list([
            dict({
              'Fn::Join': list([
                '',
                list([
                  'arn:',
                  dict({
                    'Ref': 'AWS::Partition',
                  }),
                  ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ]),
              ]),
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Role',
      }),
      'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF': dict({
        'Properties': dict({
          'PolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': list([
                  's3:GetObject*',
                  's3:GetBucket*',
                  's3:List*',
                ]),
                'Effect': 'Allow',
                'Resource': list([
                  dict({
                    'Fn::Join': list([
                      '',
                      list([
                        'arn:',
                        dict({
                          'Ref': 'AWS::Partition',
                        }),
                        ':s3:::',
                        dict({
                          'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1',
                        }),
                      ]),
                    ]),
                  }),
                  dict({
                    'Fn::Join': list([
                      '',
                      list([
                        'arn:',
                        dict({
                          'Ref': 'AWS::Partition',
                        }),
                        ':s3:::',
                        dict({
                          'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1',
                        }),
                        '/*',
                      ]),
                    ]),
                  }),
                ]),
              }),
              dict({
                'Action': list([
                  's3:GetObject*',
                  's3:GetBucket*',
                  's3:List*',
                  's3:DeleteObject*',
                  's3:PutObject',
                  's3:PutObjectLegalHold',
                  's3:PutObjectRetention',
                  's3:PutObjectTagging',
                  's3:PutObjectVersionTagging',
                  's3:Abort*',
                ]),
                'Effect': 'Allow',
                'Resource': list([
                  'arn:aws:s3:::s3-scripts-bucket-390403866377',
                  'arn:aws:s3:::s3-scripts-bucket-390403866377/*',
                ]),
              }),
            ]),
            'Version': '2012-10-17',
          }),
          'PolicyName': 'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF',
          'Roles': list([
            dict({
              'Ref': 'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265',
            }),
          ]),
        }),
        'Type': 'AWS::IAM::Policy',
      }),
      'PopulatelambdaBucketAwsCliLayer9FD2CBCF': dict({
        'Properties': dict({
          'Content': dict({
            'S3Bucket': dict({
              'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1',
            }),
          }),
          'Description': '/opt/awscli/aws',
        }),
        'Type': 'AWS::Lambda::LayerVersion',
      }),
      'PopulatelambdaBucketCustomResourceEA355E6C': dict({
        'DeletionPolicy': 'Delete',
        'Properties': dict({
          'DestinationBucketName': 's3-scripts-bucket-390403866377',
          'Prune': False,
          'ServiceToken': dict({
            'Fn::GetAtt': list([
              'CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536',
              'Arn',
            ]),
          }),
          'SourceBucketNames': list([
            dict({
              'Fn::Sub': 'cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1',
            }),
          ]),
          'SourceObjectKeys': list([
            'ff75372ec91838d1a1e0eff2df6ab6c5eac524c1d9a2ef1eb972824e543c24fd.zip',
          ]),
        }),
        'Type': 'Custom::CDKBucketDeployment',
        'UpdateReplacePolicy': 'Delete',
      }),
      'lamdataextraction': dict({
        'DeletionPolicy': 'Retain',
        'DependsOn': list([
          'PopulatelambdaBucketAwsCliLayer9FD2CBCF',
          'PopulatelambdaBucketCustomResourceEA355E6C',
        ]),
        'Properties': dict({
          'Architectures': list([
            'x86_64',
          ]),
          'Code': dict({
            'S3Bucket': 's3-scripts-bucket-390403866377',
          }),
          'Description': '',
          'EphemeralStorage': dict({
            'Size': 1024,
          }),
          'FunctionName': 'lam_data_extraction',
          'Handler': 'lambda_function.lambda_handler',
          'Layers': list([
            'arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15',
          ]),
          'MemorySize': 1024,
          'Role': 'arn:aws:iam::390403866377:role/service-role/first-test-role-vn00mm7y',
          'Runtime': 'python3.12',
          'Tags': list([
            dict({
              'Key': 'Name',
              'Value': 'lam_data_extraction',
            }),
          ]),
          'Timeout': 300,
        }),
        'Type': 'AWS::Lambda::Function',
        'UpdateReplacePolicy': 'Retain',
      }),
      'laminsertdatacatalog': dict({
        'DeletionPolicy': 'Retain',
        'DependsOn': list([
          'PopulatelambdaBucketAwsCliLayer9FD2CBCF',
          'PopulatelambdaBucketCustomResourceEA355E6C',
        ]),
        'Properties': dict({
          'Architectures': list([
            'x86_64',
          ]),
          'Code': dict({
            'S3Bucket': 's3-scripts-bucket-390403866377',
          }),
          'Description': '',
          'EphemeralStorage': dict({
            'Size': 1024,
          }),
          'FunctionName': 'lam_insert_data_catalog',
          'Handler': 'lambda_function.lambda_handler',
          'Layers': list([
            'arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15',
          ]),
          'MemorySize': 1024,
          'Role': 'arn:aws:iam::390403866377:role/service-role/first-test-role-vn00mm7y',
          'Runtime': 'python3.12',
          'Tags': list([
            dict({
              'Key': 'Name',
              'Value': 'lam_insert_data_catalog',
            }),
          ]),
          'Timeout': 300,
        }),
        'Type': 'AWS::Lambda::Function',
        'UpdateReplacePolicy': 'Retain',
      }),
      'lamtransformarprecipitaciones': dict({
        'DeletionPolicy': 'Retain',
        'DependsOn': list([
          'PopulatelambdaBucketAwsCliLayer9FD2CBCF',
          'PopulatelambdaBucketCustomResourceEA355E6C',
        ]),
        'Properties': dict({
          'Architectures': list([
            'x86_64',
          ]),
          'Code': dict({
            'S3Bucket': 's3-scripts-bucket-390403866377',
          }),
          'Description': '',
          'EphemeralStorage': dict({
            'Size': 1024,
          }),
          'FunctionName': 'lam_transformar_precipitaciones',
          'Handler': 'lambda_function.lambda_handler',
          'Layers': list([
            'arn:aws:lambda:us-east-1:336392948345:layer:AWSSDKPandas-Python312:15',
          ]),
          'MemorySize': 1024,
          'Role': 'arn:aws:iam::390403866377:role/service-role/first-test-role-vn00mm7y',
          'Runtime': 'python3.12',
          'Tags': list([
            dict({
              'Key': 'Name',
              'Value': 'lam_transformar_precipitaciones',
            }),
          ]),
          'Timeout': 300,
        }),
        'Type': 'AWS::Lambda::Function',
        'UpdateReplacePolicy': 'Retain',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
